File: packages/ik_rules/lib/src/cosmetics/cosmetics.dart
import 'package:collection/collection.dart';
import 'package:ik_content/ik_content.dart';

import '../equipment/loadout.dart';
import '../save/generated/save_models.dart';

List<CosmeticSlotRow> cosmeticSlots(GameDatabase db) => db.cosmeticSlots;

CosmeticSlotRow? cosmeticSlotById(GameDatabase db, String slotId) {
  return db.cosmeticSlots.firstWhereOrNull((row) => row.raw['Cosmetic Slot ID'] == slotId);
}

List<CosmeticRow> cosmeticsForSlot(GameDatabase db, String slotId) {
  return db.cosmetics.where((row) => row.raw['Cosmetic Slot ID'] == slotId).toList();
}

CosmeticRow? cosmeticById(GameDatabase db, String cosmeticId) {
  return db.cosmetics.firstWhereOrNull((row) => row.raw['Cosmetic ID'] == cosmeticId);
}

bool isCosmeticUnlocked(PlayerSave save, String cosmeticId) {
  return save.cosmetics.unlocked.contains(cosmeticId);
}

String? equippedCosmeticId(PlayerSave save, String slotId) {
  return save.cosmetics.equipped[slotId];
}

/// The outcome of unlocking a cosmetic, including whether it was the first one
/// ever, which the wardrobe-unlock popup keys off.
class CosmeticGrantResult {
  const CosmeticGrantResult({required this.save, required this.granted, required this.isFirstEver});

  final PlayerSave save;
  final bool granted;
  final bool isFirstEver;
}

/// Unlocks a cosmetic permanently. Owning multiples has no effect, and since
/// cosmetics live in their own always-owned collection rather than the bag,
/// granting one can never fail for lack of space.
CosmeticGrantResult grantCosmetic(PlayerSave save, String cosmeticId) {
  final current = save.cosmetics;
  if (current.unlocked.contains(cosmeticId)) {
    return CosmeticGrantResult(save: save, granted: false, isFirstEver: false);
  }
  return CosmeticGrantResult(
    save: save.copyWith(cosmetics: current.copyWith(unlocked: [...current.unlocked, cosmeticId])),
    granted: true,
    isFirstEver: current.unlocked.isEmpty,
  );
}

/// Equips an owned cosmetic into its slot, or unequips when [cosmeticId] is null.
///
/// Refused when the cosmetic is not unlocked yet or belongs in another slot.
/// Unlike gear, this can never fail for capacity reasons.
EquipResult equipCosmetic(GameDatabase db, PlayerSave save, String slotId, String? cosmeticId) {
  final current = save.cosmetics;
  if (cosmeticId == null) {
    return EquipResult.ok(
      save.copyWith(cosmetics: current.copyWith(equipped: {...current.equipped, slotId: null})),
    );
  }
  final cosmetic = cosmeticById(db, cosmeticId);
  if (cosmetic == null) return const EquipResult.failed('Unknown Cosmetic.');
  if (cosmetic.raw['Cosmetic Slot ID'] != slotId) {
    return const EquipResult.failed('That Cosmetic does not belong in this slot.');
  }
  if (!current.unlocked.contains(cosmeticId)) {
    return const EquipResult.failed('That Cosmetic has not been unlocked yet.');
  }
  return EquipResult.ok(
    save.copyWith(cosmetics: current.copyWith(equipped: {...current.equipped, slotId: cosmeticId})),
  );
}

File: src/game/cosmetics/cosmetics.ts
import type { CosmeticRow, CosmeticSlotRow, GameDatabase } from '../data/types'
import type { EquipResult } from '../equipment/loadout'
import type { CosmeticsState, PlayerSave } from '../save/types'

export function cosmeticSlots(db: GameDatabase): CosmeticSlotRow[] {
  return db.CosmeticSlots
}

export function cosmeticSlotById(db: GameDatabase, slotId: string): CosmeticSlotRow | undefined {
  return db.CosmeticSlots.find((row) => row['Cosmetic Slot ID'] === slotId)
}

export function cosmeticsForSlot(db: GameDatabase, slotId: string): CosmeticRow[] {
  return db.Cosmetics.filter((row) => row['Cosmetic Slot ID'] === slotId)
}

export function cosmeticById(db: GameDatabase, cosmeticId: string): CosmeticRow | undefined {
  return db.Cosmetics.find((row) => row['Cosmetic ID'] === cosmeticId)
}

export function cosmeticByItemId(db: GameDatabase, itemId: string): CosmeticRow | undefined {
  return db.Cosmetics.find((row) => row['Item ID'] === itemId)
}

function cosmeticsState(save: PlayerSave): CosmeticsState {
  return save.cosmetics ?? { unlocked: [], equipped: {} }
}

export function isCosmeticUnlocked(save: PlayerSave, cosmeticId: string): boolean {
  return cosmeticsState(save).unlocked.includes(cosmeticId)
}

export function equippedCosmeticId(save: PlayerSave, slotId: string): string | null {
  return cosmeticsState(save).equipped[slotId] ?? null
}

/**
 * Unlock a Cosmetic permanently (idempotent — owning multiples has no
 * effect). Cosmetics live in their own always-owned collection rather than
 * the inventory bag, so granting one can never fail for lack of bag space.
 */
export function grantCosmetic(
  save: PlayerSave,
  cosmeticId: string,
): { save: PlayerSave; granted: boolean; isFirstEver: boolean } {
  const current = cosmeticsState(save)
  if (current.unlocked.includes(cosmeticId)) {
    return { save, granted: false, isFirstEver: false }
  }
  const isFirstEver = current.unlocked.length === 0
  return {
    save: {
      ...save,
      cosmetics: {
        ...current,
        unlocked: [...current.unlocked, cosmeticId],
      },
    },
    granted: true,
    isFirstEver,
  }
}

/**
 * Equip an owned Cosmetic into its slot, or pass `cosmeticId: null` to
 * unequip. Blocked (not performed) if the Cosmetic isn't unlocked yet or
 * doesn't belong to the given slot — this can never fail for capacity
 * reasons the way gear equip can, since Cosmetics don't use bag slots.
 */
export function equipCosmetic(
  db: GameDatabase,
  save: PlayerSave,
  slotId: string,
  cosmeticId: string | null,
): EquipResult {
  const current = cosmeticsState(save)
  if (cosmeticId == null) {
    return {
      ok: true,
      save: { ...save, cosmetics: { ...current, equipped: { ...current.equipped, [slotId]: null } } },
    }
  }
  const cosmetic = cosmeticById(db, cosmeticId)
  if (!cosmetic) {
    return { ok: false, reason: 'Unknown Cosmetic.' }
  }
  if (cosmetic['Cosmetic Slot ID'] !== slotId) {
    return { ok: false, reason: 'That Cosmetic does not belong in this slot.' }
  }
  if (!current.unlocked.includes(cosmeticId)) {
    return { ok: false, reason: 'That Cosmetic has not been unlocked yet.' }
  }
  return {
    ok: true,
    save: {
      ...save,
      cosmetics: { ...current, equipped: { ...current.equipped, [slotId]: cosmeticId } },
    },
  }
}
